### ISSUE TITLE
cpufreq-applet: freq/units labels render partially or not at all

### ISSUE THREAD

**Issue by dwtokarski** [mate-desktop/mate-applets#268]
#### Expected behaviour
A cpufreq applet placed on the panel should render completely and remain so.

#### Actual behaviour
Almost always, the last column or two of pixels on the units label is missing (e.g. the z in Mhz is missing the right side).

Somewhat less often, more or all the the units label is missing entirely; likewise for the frequency label. On even more rare occasions the applet doesn't render at all, although it's really there as hovering over its known location will pop up the tooltip, left click opens the freq/governors menu, etc.

I have reason to think the total failure to render is a separate issue from the label truncations behavior.

#### Steps to reproduce the behaviour
No special action required. Just log in to a desktop which has one or more cpufreq-applets on a panel. Can also happen after resume from suspend or even any random moment.

#### MATE general version
1.18.0 currently, though this has been happening since at least 1.16.x.

#### Package version
Currently mate-applets-1.18.1-1.fc26.x86_64 but see above.

#### Linux Distribution
Fedora 25, 26

#### Link to downstream report of your Distribution
None yet.

#### Proposed Patch
The patch below addresses the partial labels issue. I've been running this for several days and it seems solid.

The problem seems to have been in cpufreq-applet.c, where the get_max_???_width routines were trying to find the pixel width of various strings by inserting text into a freshly allocated GtkLabel and calling gtk_widget_get_preferred_width() on it, which just returns zero results. Wild guess: the freestanding GtkLabel has no associated font, so the width can't be determined.  At some point someone decided it would be better to get the widths of the applet-> label instead but they're not initialized at first and who knows where the string pointer in that object was pointing. Wildly different values were returned from time to time, and the first time a positive value is acquired it becomes the sticky width for that label in *applet.

The patch gets around this by placing string values in the labels of the applet, then getting the preferred widths. 

```
diff -ur ../mate-applets-1.18.1-ORIG/cpufreq/src/cpufreq-applet.c ./cpufreq/src/cpufreq-applet.c
 --- ../mate-applets-1.18.1-ORIG/cpufreq/src/cpufreq-applet.c   2017-04-06 15:02:40.000000000 -0400
 +++ ./cpufreq/src/cpufreq-applet.c     2017-08-20 16:05:37.056044732 -0400
 @@ -292,7 +292,6 @@
  cpufreq_applet_get_max_label_width (CPUFreqApplet *applet)
  {
        GList *available_freqs;
 -      gint   width = 0;
  
        if (applet->max_label_width > 0)
                return applet->max_label_width;
 @@ -302,7 +301,6 @@
  
        available_freqs = cpufreq_monitor_get_available_frequencies (applet->monitor);
        while (available_freqs) {
 -              GtkWidget     *label;
                gint           label_width;
                const gchar   *text;
                gchar         *freq_text;
 @@ -312,34 +310,26 @@
                freq = atoi (text);
  
                freq_text = cpufreq_utils_get_frequency_label (freq);
 -              label = gtk_label_new (freq_text);
 +              gtk_label_set_text (GTK_LABEL (applet->label), freq_text);
                gtk_widget_get_preferred_width (applet->label, &label_width, NULL);
 -              width = MAX (width, label_width); 
 +              applet->max_label_width = MAX (applet->max_label_width, label_width);
  
                g_free (freq_text);
 -              gtk_widget_destroy (label);
  
                available_freqs = g_list_next (available_freqs);
        }
  
 -      applet->max_label_width = width;
 -
 -      return width;
 +      return applet->max_label_width;
  }
  
  static gint
  cpufreq_applet_get_max_perc_width (CPUFreqApplet *applet)
  {
 -      GtkWidget      *label;
 -      gint            width;
 -
        if (applet->max_perc_width > 0)
                return applet->max_perc_width;
  
 -      label = gtk_label_new ("100%");
 -      gtk_widget_get_preferred_width (applet->label, &width,&width);
 -      applet->max_perc_width = width +20; /*for some reason width always comes up 2 characters*/
 -      gtk_widget_destroy (label);
 +      gtk_label_set_text (GTK_LABEL (applet->label), "100%");
 +      gtk_widget_get_preferred_width (applet->label, &applet->max_perc_width, NULL);
  
        return applet->max_perc_width;
  }
 @@ -347,20 +337,19 @@
  static gint
  cpufreq_applet_get_max_unit_width (CPUFreqApplet *applet)
  {
 -      GtkWidget     *label;
 -      gint           w1, w2;
 +      gint           mwidth;
  
        if (applet->max_unit_width > 0)
                return applet->max_unit_width;
  
 -      label = gtk_label_new ("GHz");
 -      gtk_widget_get_preferred_width (applet->label, &w1, NULL);
 +      gtk_label_set_text (GTK_LABEL (applet->unit_label), "GHz");
 +      gtk_widget_get_preferred_width (applet->unit_label, &applet->max_unit_width, NULL);
  
 -      gtk_label_set_text (GTK_LABEL (label), "MHz");
 -      gtk_widget_get_preferred_width (applet->label, &w2, NULL);
 +      gtk_label_set_text (GTK_LABEL (applet->unit_label), "MHz");
 +      gtk_widget_get_preferred_width (applet->unit_label, &mwidth, NULL);
 +
 +      applet->max_unit_width = MAX (applet->max_unit_width, mwidth);
  
 -      gtk_widget_destroy (label);
 -      applet->max_unit_width = MAX (w1, w2)-1;
        return applet->max_unit_width;
  }
  
 @@ -757,6 +746,20 @@
          guint        cpu;
          const gchar *governor;
  
 +        /* Call refresh only the first time.
 +         *
 +         * Because the label width determinations now modify the actual labels
 +         * in *applet to size the max label width in pixels, do this first
 +         * before setting the label text below.
 +         */
 +        if (applet->need_refresh) {
 +                gint w;
 +
 +                cpufreq_applet_get_preferred_width(applet, &w, &w);
 +                cpufreq_applet_refresh (applet);
 +                applet->need_refresh = FALSE;
 +        }
 +
          cpu = cpufreq_monitor_get_cpu (monitor);
          freq = cpufreq_monitor_get_frequency (monitor);
          perc = cpufreq_monitor_get_percentage (monitor);
 @@ -810,12 +813,6 @@
                gtk_widget_set_tooltip_text (GTK_WIDGET (applet), text_tip);
                g_free (text_tip);
        }
 -
 -        /* Call refresh only the first time */
 -        if (applet->need_refresh) {
 -                cpufreq_applet_refresh (applet);
 -              applet->need_refresh = FALSE;
 -        }
  }
  
  static gint
~
```

**Comment by flexiondotorg** [comment]
Please can you submit your patch as a pull request ☺️

**Comment by dwtokarski** [comment]
Sure, not a problem, but it'll take me a bit to set it up.

Should I base off of v1.18.1, or would you prefer something else?

**Comment by raveit65** [comment]
To master please.

**Comment by lukefromdc** [comment]
I just tested this, found it did not always work. Usually I do not get this bug, both with master and this code I tried changing between various themes, which sometimes would invoke it in BlackMATE. Once and once only, on restarting the panel after changing back to my own theme I get a truncated z in GHz. 
![applet_w_cutoff_units](https://user-images.githubusercontent.com/14036792/29854249-b95abf12-8d12-11e7-80f7-543ec77506b2.png)

My theme uses bold text, which caused the applet to jump until I added this to it:

```
/*Stop jumping cpufreq labels with bold text*/
#PanelApplet>widget>box>box>label{
	min-width: 35px;
}

Both with and without this I had had issues in the past with truncating the z on changing back to my theme, and sometimes on machines and kernels that would initially return "0GHZ" as value I would get a truncated applet. That was worst on one with an 800MHZ idle speed, and the same fix as for the jumping with bold test put a stop to that truncation. Now I saw a single, isolated instance of a truncated label even with the minimum width set, not sure how that was possible.
```

**Comment by raveit65** [comment]
@dwtokarski 
With which theme(s) you saw the issue?
...i will try to reproduce it.

**Comment by dwtokarski** [comment]
@raveit65 
  BlueMenta controls with Bluebuntu window borders, and Bluecurve icons and pointer.

  Because of the way the applet holds onto the first positive width value it gets for the labels, you can best provoke the problem by adding cpufreq applets to the panel. If/when one turns out to be truncated, any others already present should be unaffected. Killing the panel works too.

@lukefromdc 
  Your observations are to be expected. The label widths are determined only once when the applet is instantiated, presumably using whatever font is currently configured, and the pixel width stored. Thereafter the stored value would be used, although after some debugging today it appears those widths are never queried again anyway, even when the font is changed.  Generally, I'd expect anything you do which makes the font larger--theme change, type face, font size--than when the applet is instantiated could cause truncation.

So today I tried removing the check for an existing positive width from all the cpufreq_applet_get_max_XXXX_width() procedures. I also removed the check for applet->need_refresh flag in cpufreq_applet_update().  What I expected that to do was force a fresh determination of label widths each time the label contents changed, including a font change. That and forcing a cpufreq_applet_refresh() call each time would, I thought, render the labels without truncation.  But no, I just see what @lukefromdc describes.  Oh, and somewhat (un?)related, I find it truly bizarre that nothing on the panel renders any italic font.

My problem is I don't know enough yet about gtk to understand what has to happen to make updated widget attributes render visibly.  More debugging is in order.

You might want to put off following through with that pull until this is sorted out.  It really does seem like we're close.

**Comment by raveit65** [comment]
Ok, Menta themes are our default and i like to use it. :-)
So i can test this several days to reproducing the issue.
....with applets for 4 cores.

**Comment by monsta** [comment]
I can't reproduce it, but I found https://github.com/mate-desktop/mate-applets/commit/300ef94527f603311fed18d7d632740ecf56163c which might be related.
So it always calls `gtk_widget_get_preferred_width` on `applet->label`, while a local label variable is just created and destroyed, but not used.

I see in https://github.com/mate-desktop/mate-applets/pull/269 it's changed to modify `applet->label` instead. Are `cpufreq_applet_get_max_xxx_width` functions supposed to have a side effect?

**Comment by lukefromdc** [comment]
We could fix this by hardocding a sane minumum size, but letting the labels be set larger when a theme requires it (say to accomodate a large or bold font).  I had originally hardcoded in my summer 2015 kludge to make the labels show up at all because the original function to get the width that worked with GTK2 was always returning zero width in GTK3.

**Comment by monsta** [comment]
Well, that sucks. I just tested it and `gtk_widget_get_preferred_width` indeed returns zero when called on a widget that's not visible (which is true for those local label variables).

**Comment by monsta** [comment]
Wonder what's going on in gnome-flashback/gnome-panel. They still have that code (see [here](https://github.com/GNOME/gnome-applets/blob/master/cpufreq/src/cpufreq-applet.c)).

**Comment by monsta** [comment]
Ok, I've found an alternative way to get the width of an invisible label. Check https://github.com/mate-desktop/mate-applets/pull/272 guys.

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/mate-applets/pull/272 did not work, gave me the worst truncation problems ever

**Comment by monsta** [comment]
Added possible fix to https://github.com/mate-desktop/mate-applets/pull/272 in a new commit.

Also added https://github.com/mate-desktop/mate-applets/pull/274 to fix truncation issue when you increase font size in mate-appearance-properties. It's a separate issue, and the fix can be applied separately as well.

**Comment by lukefromdc** [comment]
We've just has another user report of this
https://github.com/mate-desktop/mate-panel/issues/658
Closed it as a duplicate and against the wrong package. Since we seem to have a working fix we should merge it and not let this one fall by the wayside.

**Comment by monsta** [comment]
Fix is released in version 1.18.2.

**Comment by lukefromdc** [comment]
In working on https://github.com/mate-desktop/mate-applets/pull/280 , I think I found the reason this applet truncated rather than jumping on wider text: something seems to set a maximum width for it based on the font-size selected, and a minimum size set in a theme can exceed this. 

It is still possible to demonstrate truncation with a large minimum size in a theme, but truncatiion on theme changes or bold text (say with a switch between 800 MHZ and some number of GHZ) is gone thanks to this being merged.
